### ISSUE TITLE
Migrate to math/rand/v2

### ISSUE THREAD

**Issue by rueian** [redis/rueidis#552]
The [math/rand/v2](https://pkg.go.dev/math/rand/v2) has several important improvements, such as eliminating the need to wrap rand.NewSource with sync.Pool in these places: https://github.com/search?q=repo%3Aredis%2Frueidis+rand.NewSource&type=code

<img width="1373" alt="image" src="https://github.com/redis/rueidis/assets/2727535/a3b69e55-31c4-44f5-a7d6-6a884e0c82a0">

We should eventually remove all the above `rand.NewSource + sync.Pool` usages and replace all the usages of `math/rand` with `math/rand/v2`.

However `math/rand/v2` is only available in Go 1.22, we need to use the `build tag` to support both old and new usages.

**Comment by SoulPancake** [comment]
Can I take this up @rueian

**Comment by rueian** [comment]
> Can I take this up @rueian

Sure. Since this task is a little more complicated than previous ones, could you explain what you are going to do before making any changes?

**Comment by SoulPancake** [comment]
With regards to not requiring to re-use using sync.Pool I was assuming we could do something like this

```
package om

import (
	"io"
	"math/rand/v2"
	"time"

	"github.com/oklog/ulid/v2"
)

type Src int

func (s Src) Uint64() uint64 {
	return uint64(s)
}

type ioReadRnd rand.Rand

func (i ioReadRnd) Read(p []byte) (n int, err error) {
	return io.ReadFull(i, p)
}

func id() string {
	src := Src(time.Now().UnixNano())
	newRnd := ioReadRnd(*rand.New(src))
	var entropies = ulid.Monotonic(newRnd, 0)
	n := time.Now()
	id := ulid.MustNew(ulid.Timestamp(n), entropies)
	return id.String()
}

```

**Comment by SoulPancake** [comment]
I was attempting something
https://go.dev/blog/randv2#:~:text=The%20Read%20Mistake,remove%20Read%20entirely.

But I am still thinking about the workaround for this

I assumed we wouldn't use sync.Pool because we no longer need to re-use
package om

import (
	"math/rand/v2"
	"time"

	"github.com/oklog/ulid/v2"
)

type Src int

func (s Src) Uint64() uint64 {
	return uint64(s)
}

func id() string {
	src := Src(time.Now().UnixNano())
	newRnd := rand.New(src)
	var entropies = ulid.Monotonic(newRnd, 0)
	n := time.Now()
	id := ulid.MustNew(ulid.Timestamp(n), entropies)
	return id.String()
}
@rueian

**Comment by SoulPancake** [comment]
Perhaps something like this can work @rueian 
obviously leaving aside the build tag 

```
package om

import (
	"math/rand/v2"
	"time"

	"github.com/oklog/ulid/v2"
)

type Src int64

func (s Src) Uint64() uint64 {
	return uint64(s)
}

type ioReadRnd struct {
	r *rand.Rand
}

func (i ioReadRnd) Read(p []byte) (n int, err error) {
	for len(p) > 0 {
		val := i.r.Uint64()
		for i := 0; i < 8 && len(p) > 0; i++ {
			p[0] = byte(val)
			val >>= 8
			p = p[1:]
			n++
		}
	}
	return n, nil
}

func id() string {
	src := Src(time.Now().UnixNano())
	rnd := rand.New(src)
	newRnd := ioReadRnd{r: rnd}
	entropy := ulid.Monotonic(newRnd, 0)
	n := time.Now()
	id := ulid.MustNew(ulid.Timestamp(n), entropy)
	return id.String()
}
```

**Comment by JayChavan-23** [comment]
Hey @SoulPancake do you wanna work on this issue together?

**Comment by rueian** [comment]
> I was attempting something https://go.dev/blog/randv2#:~:text=The%20Read%20Mistake,remove%20Read%20entirely.
> 
> But I am still thinking about the workaround for this
> 
> I assumed we wouldn't use sync.Pool because we no longer need to re-use package om
> 
> ```
> import (
> 	"math/rand/v2"
> 	"time"
> 
> 	"github.com/oklog/ulid/v2"
> )
> 
> type Src int
> 
> func (s Src) Uint64() uint64 {
> 	return uint64(s)
> }
> 
> func id() string {
> 	src := Src(time.Now().UnixNano())
> 	newRnd := rand.New(src)
> 	var entropies = ulid.Monotonic(newRnd, 0)
> 	n := time.Now()
> 	id := ulid.MustNew(ulid.Timestamp(n), entropies)
> 	return id.String()
> }
> ```
> 
> @rueian

Hi @SoulPancake, I think we can just replace the whole function with `ulid.Make().String()`.

**Comment by SoulPancake** [comment]
So, the entropy will be straight up from the Make()  which takes in the current timestamp
So, Essentially we wouldn't need math/rand/v2 for the new build tag? @rueian

**Comment by rueian** [comment]
For the `ulid.go`, yes, we can actually just remove the entire file.

**Comment by SoulPancake** [comment]
>replace all the usages of math/rand with math/rand/v2.

Do we use build tags for literally all the usages @rueian 
even those except the ones you had mentioned above? 
For example

```
cluster.go
rueidis.go
sentinel.go
```

**Comment by rueian** [comment]
Hi @SoulPancake,

We do need to use build tags for all the `math/rand` usages, but we need to centralize them to one file first.

For example, we can have random functions defined in `rand.go` and `rand.1.22.go` files and use them in other files. So that we don't have to duplicate most of the files which use `math/rand`.

Here is another example: https://github.com/maypok86/otter/tree/main/internal/xruntime

**Comment by SoulPancake** [comment]
We can create two random utils within utils with build tags and use it everywhere @rueian

**Comment by rueian** [comment]
That is a good idea. Just keep in mind that we should not expose the utils as public functions to users.

**Comment by rueian** [comment]
Hi @JayChavan-23,

@SoulPancake has already done most of the migration but we haven't tested it with go1.22 on Github Actions. Would you like to work on that part? Here is the issue for that: https://github.com/redis/rueidis/issues/554

**Comment by JayChavan-23** [comment]
Yes thankyou @rueian I'll work on it, Cheers.

**Comment by rueian** [comment]
Closed by #553.
